content/blog/2018/06/2018-06-26-new-api-token-system.adoc

```diff
@@ -0,0 +1,111 @@
+
+An API token is an authentication mechanism that allows a tool (script, application, etc.) to impersonate the user
+in order to obtain the same permissions. 
+This is especially useful when you do not want to provide your password in plain text in script.
+They are not meant to replace the regular authentication when visiting the application with a browser.
+
+There were two problems with the existing API token system. First, 
+reported in link:https://issues.jenkins-ci.org/browse/JENKINS-32442[JENKINS-32442], when a user is created, 
+the application automatically generates an API token for them by default. 
+This behavior increases the attack surface, in other words, 
+the users must protect their application's password and additionally the API token because they have both 
+the same value in terms of security. 
+Often the users are not even aware of that fact and so will never use this token and worse will never regenerate a new one
+(which will have the consequence of revoking the current one).
+
+The second problem was reported in link:https://issues.jenkins-ci.org/browse/JENKINS-32776[JENKINS-32776]: the tokens are stored 
+on disk in an encrypted version. 
+In terms of security this means that if you have the key, you could retrieve the plaintext 
+(i.e. the token that could be put in a request to be authenticated as the user). 
+An attacker that has access to a backup of your instance will be able to take advantage of that recoverable format
+to retrieve the token and use it to impersonate the user without leaving any suspect traces (completely legitimate connection), 
+like when your password is stolen.
+
+### New approach
+
+The main objective of this new system is to provide tokens that are stored in an unidirectional way on the disk,
+i.e. using a hashing algorithm (in this particular case SHA-256).
+
+First the main drawback, you will not be able to recover your tokens. 
+This only affects the tokens generated with the new system.
+All the legacy tokens are still recoverable.
+
+To mitigate this drawback completely and enforce better practices, we provide different new features with this version:
+
+* You can have multiple active API tokens at the same time, compared to the single one from before.
+
+* You can name your tokens to know where they are used (and rename them after creation if desired). 
+We advise to name the token according to the application / script where it will be used.
+
+* You can track the usage of your tokens.
+Every token keeps a record of the number of uses and the date of the last use.
+This will allow you to better know which tokens are really used and which are no longer actively required.
+When a token was created more than 6 months ago, we display the creation date in orange and when it's more than 1 year, 
+the date is in red.
+The goal is to remind the user that tokens are more secure when you regenerate them often.
+The longer a token lives, the greater the chance it's going to be stolen.
```

Comment:
IIRC the password rotation is a problem due to the way it is "solved" by users like the users that have set their password to be dependant on the rotation frequency: Must be changed every month, so I put a "-jan", "-feb", "-mar" in suffix.

For token it's the same as for certificates, the good practice is to change them regularly to avoid situation where a certificate was stolen / in the wild but the administrator is not aware of that. Another point is when an attacker found an old certificate like in backup or legacy application, it's better if that certificate is no more valid thanks to the rotations."